components/tidb_query/src/expr_util/collation/utf8mb4.rs

```diff
@@ -58,17 +58,277 @@ const GENERAL_CI_PLANE_01: [u16; 256] = [
+    0xFFCC, 0xFFCD, 0xFFCE, 0xFFCF, 0xFFD0, 0xFFD1, 0xFFD2, 0xFFD3, 0xFFD4, 0xFFD5, 0xFFD6, 0xFFD7,
+    0xFFD8, 0xFFD9, 0xFFDA, 0xFFDB, 0xFFDC, 0xFFDD, 0xFFDE, 0xFFDF, 0xFFE0, 0xFFE1, 0xFFE2, 0xFFE3,
+    0xFFE4, 0xFFE5, 0xFFE6, 0xFFE7, 0xFFE8, 0xFFE9, 0xFFEA, 0xFFEB, 0xFFEC, 0xFFED, 0xFFEE, 0xFFEF,
+    0xFFF0, 0xFFF1, 0xFFF2, 0xFFF3, 0xFFF4, 0xFFF5, 0xFFF6, 0xFFF7, 0xFFF8, 0xFFF9, 0xFFFA, 0xFFFB,
+    0xFFFC, 0xFFFD, 0xFFFE, 0xFFFF,
+];
+
+#[rustfmt::skip]
+const GENERAL_CI_PLANE_TABLE: [Option<&[u16]>; 256] = [
+    Some(&GENERAL_CI_PLANE_00),
+    Some(&GENERAL_CI_PLANE_01),
+    Some(&GENERAL_CI_PLANE_02),
+    Some(&GENERAL_CI_PLANE_03),
+    Some(&GENERAL_CI_PLANE_04),
+    Some(&GENERAL_CI_PLANE_05),
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 
+    None, None, None, None, None, None, None, None,
+    Some(&GENERAL_CI_PLANE_1E),
+    Some(&GENERAL_CI_PLANE_1F),
+    None,
+    Some(&GENERAL_CI_PLANE_21),
+    None, None,
+    Some(&GENERAL_CI_PLANE_24),
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+    None, None, None, None, None, None, None, None, None, None,
+    Some(&GENERAL_CI_PLANE_FF),
+];
+
 pub struct CollatorUtf8Mb4GeneralCi;
 
 #[inline]
-fn general_ci_weight(c: char) -> u16 {
-    let r = c as u32;
-    let plane = if (r >> 8) > 0 {
-        &GENERAL_CI_PLANE_01
+fn general_ci_convert(c: char) -> u16 {
+    let r = c as usize;
+    if r > 0xFFFF {
+        return 0xFFFD;
+    }
+    if let Some(plane) = GENERAL_CI_PLANE_TABLE[r >> 8] {
+        plane[r & 0xFF]
     } else {
-        &GENERAL_CI_PLANE_00
-    };
-    plane[(r & 0xFF) as usize]
+        r as u16
```

Comment:
Is wrapping desired behaviour in this case? Do we need to use `wrapping_xxx` function? (e.g. for 3 bytes chars, the output is only 2 bytes)

Reply:
Wrapping will not happen. `r` is checked at L324